### PR TITLE
Creates Signup Screen UI

### PR DESCRIPTION
**Pull request by safaorhan**
## Background
This PR adds .... which are mentioned in the task...

## Changes
- Added Validator classes to encapsulate validation logic.
- ...

## Screenshots
...

### CODE REVIEW

File: app/src/main/java/com/example/cactusnotes/SignUpActivity.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import com.example.cactusnotes.validations.UsernameValidation
+
+
+class SignUpActivity : AppCompatActivity() {
+    private lateinit var binding: SignUpActivityBinding
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = SignUpActivityBinding.inflate(layoutInflater)
+        setContentView(binding.root)
+        this.setTitle("Sign Up")
```
Comment (safaorhan) [inline app/src/main/java/com/example/cactusnotes/SignUpActivity.kt:18]: 1. User visible strings should always be extracted into strings.xml.
2. As far as I know the "recommended and widely used" way of setting title is `supportActionBar?.title = R.string.your_title`
3. If you decide to use Activity#setTitle use property access syntax instead and also drop the keyword "**this**": `title = R.string.your_title`

File: app/src/main/java/com/example/cactusnotes/SignUpActivity.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+                Toast.makeText(applicationContext, R.string.successful, Toast.LENGTH_LONG).show()
+            }
+        }
+    }
+
+    private fun isEmailValid() = binding.emailEditText.isFieldValid(EmailValidator())
+
+    private fun isPasswordValid() = binding.passwordEditText.isFieldValid(PasswordValidator())
+
+    private fun isUserNameValidation() = binding.userNameEditText.isFieldValid(UsernameValidation())
```
Comment (safaorhan) [inline app/src/main/java/com/example/cactusnotes/SignUpActivity.kt:31]: isUserNameValid()

File: app/src/main/java/com/example/cactusnotes/ValidaitonsFunctions.kt [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.example.cactusnotes
+
+fun String.hasSpecialLetters() = any { !it.isLetterOrDigit() }
+
+fun String.isNameValid() = all { it.isDigit() || it.isLowerCase() || it == '_' }
+
+fun String.hasDigit() = any { it.isDigit() }
```
Comment (safaorhan) [inline app/src/main/java/com/example/cactusnotes/ValidaitonsFunctions.kt:7]: hasDigits()*

File: app/src/main/java/com/example/cactusnotes/validations/EmailValidator.kt [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package com.example.cactusnotes.validations
+
+import com.example.cactusnotes.ValidationResult
+
+class EmailValidator() : Validator {
+    private fun String.emailValid() = contains("@") && contains(".")
```
Comment (safaorhan) [inline app/src/main/java/com/example/cactusnotes/validations/EmailValidator.kt:6]: isEmailValid.

This line could be inlined.

File: app/src/main/java/com/example/cactusnotes/validations/EmailValidator.kt [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    override fun value(field: String): ValidationResult {
+        return if (field.isEmpty()) {
+            ValidationResult(false, "Email is required.")
```
Comment (safaorhan) [inline app/src/main/java/com/example/cactusnotes/validations/EmailValidator.kt:9]: User visible strings should always be extracted into strings.xml

File: app/src/main/java/com/example/cactusnotes/validations/UsernameValidation.kt [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package com.example.cactusnotes.validations
+
+import com.example.cactusnotes.ValidationResult
+import com.example.cactusnotes.isNameValid
+
+class UsernameValidation() : Validator {
```
Comment (safaorhan) [inline app/src/main/java/com/example/cactusnotes/validations/UsernameValidation.kt:6]: `UsernameValidator` would be a better name

File: app/src/main/res/layout/sign_up_activity.xml [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    tools:context=".SignUpActivity">
+
+    <com.google.android.material.textfield.TextInputLayout
+        android:id="@+id/emailEditText"
+        style="@style/Widget.MaterialComponents.TextInputLayout.OutlinedBox"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_gravity="center"
+        android:layout_marginStart="18dp"
+        android:layout_marginTop="18dp"
```
Comment (safaorhan) [inline app/src/main/res/layout/sign_up_activity.xml:18]: This margin top should be reduced / removed to match with the design in Figma

File: app/src/main/res/layout/sign_up_activity.xml [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    android:orientation="vertical"
+    tools:context=".SignUpActivity">
+
+    <com.google.android.material.textfield.TextInputLayout
+        android:id="@+id/emailEditText"
+        style="@style/Widget.MaterialComponents.TextInputLayout.OutlinedBox"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_gravity="center"
+        android:layout_marginStart="18dp"
```
Comment (safaorhan) [inline app/src/main/res/layout/sign_up_activity.xml:17]: 18dp spacings should be 16dp when I check the design

File: app/src/main/res/layout/sign_up_activity.xml [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        android:layout_marginTop="18dp"
+        android:layout_marginEnd="18dp"
+        android:hint="@string/password"
+        app:endIconMode="password_toggle"
+        app:errorEnabled="false"
+        app:layout_constraintTop_toBottomOf="@+id/userNameEditText">
```
Comment (safaorhan) [inline app/src/main/res/layout/sign_up_activity.xml:59]: Unused attribute. Should be removed.

File: app/src/main/res/layout/sign_up_activity.xml [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    <Button
+        android:id="@+id/sign_up_button"
+        android:layout_width="match_parent"
+        android:layout_height="65dp"
+        android:layout_gravity="center"
+        android:layout_marginStart="16dp"
+        android:layout_marginTop="24dp"
+        android:layout_marginEnd="16dp"
+        android:text="@string/sign_up"
+        app:layout_constraintTop_toBottomOf="@+id/passwordEditText" />
```
Comment (safaorhan) [inline app/src/main/res/layout/sign_up_activity.xml:76]: Unused attribute. Should be removed.

File: app/src/main/res/layout/sign_up_activity.xml [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        android:id="@+id/outlinedButton"
+        style="@style/Widget.MaterialComponents.Button.TextButton.Icon"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_gravity="center"
+        android:layout_marginStart="18dp"
+        android:layout_marginTop="18dp"
+        android:layout_marginEnd="18dp"
+        android:text="@string/already_have_an_account"
+        app:layout_constraintStart_toStartOf="parent"
```
Comment (safaorhan) [inline app/src/main/res/layout/sign_up_activity.xml:88]: Unused attribute. Should be removed.

File: app/src/main/res/layout/sign_up_activity.xml [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        app:layout_constraintTop_toBottomOf="@+id/sign_up_button" />
```
Comment (safaorhan) [inline app/src/main/res/layout/sign_up_activity.xml:89]: Unused attribute. Should be removed.

**Review by safaorhan — COMMENTED** [review]
1. Ids should follow lowerCamelCase naming.
2. All other XML entities (strings, colors, etc.) should follow snake_case naming.
3. Fix typos.
4. ❌ UserName ✅  Username
5. In classes with no constructor, you can delete the empty pair of parentheses.
6. Button -> MaterialButton
7. In activity_main.xml remove unused gravity attributes.
8. 65dp -> 64dp
9. ValidateInterface.kt -> Validator.kt
10. Double check filenames.
11. Validator.value(...) -> Validator.validate(...)
12:

 ```
   override fun value(field: String) = when {
        field.isEmpty() -> ValidationResult(false, R.string.required)
        field.isEmailValid().not() -> ValidationResult(false, R.string.emailInvalid)
        field.length < 5 -> ValidationResult(false, R.string.emailInvalid)
        field.length > 50 -> ValidationResult(false, R.string.emailInvalid)
        else -> ValidationResult(true, null)
    }
```

13. `private fun String.containsAtAndDot() = contains("@") && contains(".")`
14. Spread the functions inside ValidationFunctions.kt into the releavant classes.

File: app/build.gradle [hunk 13]
```diff
@@ -41,4 +44,5 @@ dependencies {
     testImplementation 'junit:junit:4.+'
     androidTestImplementation 'androidx.test.ext:junit:1.1.2'
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.3.0'
+    implementation 'com.google.android.material:material:1.3.0'
```
Comment (safaorhan) [inline app/build.gradle:47]: This dependency already exists at line 42.

File: app/src/main/java/com/example/cactusnotes/validations/PasswordValidator.kt [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+    private fun String.hasDigits() = any { it.isDigit() }
+
+    private fun String.hasUpperCaseLetters() = any { it.isUpperCase() }
+
+    private fun String.hasLowerCaseLetters() = any { it.isLowerCase() }
+
+    private fun String.hasSpecialLetters() = any { !it.isLetterOrDigit() }
+
+    private fun String.containsRequiredCharCharacters() = hasDigits() && hasLowerCaseLetters()
```
Comment (safaorhan) [inline app/src/main/java/com/example/cactusnotes/validations/PasswordValidator.kt:22]: typo in function name:

❌ containsRequiredCharCharacters
✅ containsRequiredCharacters

File: app/src/main/java/com/example/cactusnotes/validations/UsernameValidator.kt [hunk 15]
```diff
@@ -0,0 +1,15 @@
+package com.example.cactusnotes.validations
+
+import com.example.cactusnotes.R
+import com.example.cactusnotes.ValidationResult
+
+class UsernameValidator : Validator {
+    override fun validate(field: String)= when {
```
Comment (safaorhan) [inline app/src/main/java/com/example/cactusnotes/validations/UsernameValidator.kt:7]: This file is not properly formatted.

You can check *Reformat code* on commit settings to enable auto-format before each commit.

File: app/src/main/java/com/example/cactusnotes/validations/UsernameValidator.kt [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+            field.isEmpty() -> { ValidationResult(false, R.string.user_name_required) }
+            field.length < 2 -> { ValidationResult(false, R.string.user_name_sort) }
+            field.length > 20 -> { ValidationResult(false, R.string.user_name_long) }
+            !field.isNameValid() -> { ValidationResult(false, R.string.is_valid_error) }
+            else -> { ValidationResult(true, null) }
+        }
+    private fun String.isNameValid() = all { it.isDigit() || it.isLowerCase() || it == '_' }
```
Comment (safaorhan) [inline app/src/main/java/com/example/cactusnotes/validations/UsernameValidator.kt:14]: Better name for `isNameValid()` function:

`containsOnlyAllowedCharacters()`

**Review by safaorhan — COMMENTED** [review]
Huge improvement 👏👏👏 

Only a few comments, after you address those, we are 🤏 this close to merging this 😄

**Review by safaorhan — COMMENTED** [review]
After this long discussion, it's relieving to see we are coming to an end.

After somebody QAs this, we can merge it!

**Comment by muammerakca10** [comment]
Congrulations, nice job. But Password box giving only "Password is too short" error

**Comment by muammerakca10** [comment]
Congrulations, nice job. QA passed. Working perfect.

**Pull request merged**